src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/EventPipeSession.cs

```diff
@@ -84,10 +95,26 @@ public async Task StopAsync(CancellationToken cancellationToken)
         private static IpcMessage CreateStartMessage(EventPipeSessionConfiguration config)
```

Comment:
I think this PR contains two distinct sets of changes:
- Add functionality to specify rundown keywords.
- Add generic retry functionality (e.g. `RetryStrategy`).

I think this PR should focus on the first and separate out the second in a separate change. If `RetryStrategy` is removed, this PR becomes much more simplified. In it's play, something similar to https://github.com/dotnet/diagnostics/blob/main/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/DiagnosticsClient.cs#L196 can be used. I would argue that implementing a similar algorithm for starting an event pipe session will have more success in actually starting a session and it removes the complexity of the caller having to determine a proper retry strategy.

The algorithm would be something like:
- If a non-default rundown keyword is specified, then use CollectTracing4 command. If it fails for anything other than an unknown command, then throw. Unknown command should be handled and fallback to the next lower command version.
- If stack walk was requested, use CollectTracing3 command. If it fails for anything other than an unknown command, then throw. Unknown command should be handled and fallback to the next lower command version.
- Use CollectTracing2 and let it throw or complete naturally.

After all that is completed, if the need to have some type of general retry understanding is necessary, it can be designed and considered separate from adding the rundown keyword.

Reply:
That was my first try, pushing down the retry logic into `DiagnosticsClient` itself. (See the first commit in this PR). The key challenge is that the right thing to do for the `GCCollect` profile is to drop the rundown (so it stays lightweight), but it is a risky thing to do in general for all use of the `DiagnosticsClient`. 

That's why the retry logic was bundled this PR.

Reply:
+1 to Andrew's comment. I do not believe the DiagnosticClient has sufficient knowledge to know whether retrying is appropriate, and if so whether to set isRundown=true or isRundown=false in that retry.


> I think this PR should focus on the first and separate out the second in a separate change

Adding custom rundown without retry means it is only usable on .NET 9 runtimes. Since we don't want to break the scenario that already works on .NET 8 and below that means the first PR on its own would be unusable. I don't see much value in splitting the work into two PRs if the first PR would be unusable without the 2nd.

Reply:
Under what circumstances would changing the rundown keyword from a non-zero value on the initial attempt then yield a successful attempt when it was downgraded to 0 while using the same command version? The channel stating the command is unsupported shouldn't be fixable by modifying the runtime keyword field and issuing the same command version; if it is, then I think the "unsupported command" notion has been overloaded to the point of not being meaningful for taking another course of action.

Reply:
If `EventPipeSession.CreateStartMessage` is already doing the downgrade before even attempting to issue the command, then why does `EventPipeStreamProvider.ProcessEvents` have to consider this at all?

At worst, this retry logic or fallback logic should be encapsulated in `EventPipeSession.Start`/`EventPipeSession.Start`. It can be made to understand the fallback mechanism and not have the callers nor the `MonitoringSourceConfiguration` have to consider it.

Reply:
The problem is "what" to retry is some logic the `DiagnosticsClient` cannot decide on its own.

In the gc-collect case, the custom keyword provided some additional information, but it is okay if we don't it, so the right thing to do for the gc-collect case is to set `requestRundown` to `false` when we retry.

In the other scenario, say, gc-collect + cpu, you wouldn't want to set `requestRundown` to `false` if the first try using the custom keyword fails. Without the rundown using the default rundown keyword, the CPU trace is useless because you can't resolve CPU samples to managed functions. In that case you should use `requestRundown` to `true` when we retry.

That is the key issue why `DiagnosticsClient` cannot completely decide on its own "what" to retry, the caller has to be part of the story.

Reply:
> If EventPipeSession.CreateStartMessage is already doing the downgrade before even attempting to issue the command, then why does EventPipeStreamProvider.ProcessEvents have to consider this at all?

CreateStartMessage only downgrades when doing so will not change the behavior.
CollectTracing4 with RundownKeyword = 0 is identical to CollectTracing3 Rundown=false. CollectTracing4 with RundownKeyword=0x80020139 is identical to CollectTracing3 Rundown=true.
All other values of RundownKeyword have no CollectTracing3 equivalent with identical behavior. These are the cases where the developer needs to make a scenario specific choice about what they want to happen which what the RetryStrategy is specifying.

Reply:
> > If EventPipeSession.CreateStartMessage is already doing the downgrade before even attempting to issue the command, then why does EventPipeStreamProvider.ProcessEvents have to consider this at all?
> 
> CreateStartMessage only downgrades when doing so will not change the behavior. CollectTracing4 with RundownKeyword = 0 is identical to CollectTracing3 Rundown=false. CollectTracing4 with RundownKeyword=0x80020139 is identical to CollectTracing3 Rundown=true. All other values of RundownKeyword have no CollectTracing3 equivalent with identical behavior. These are the cases where the developer needs to make a scenario specific choice about what they want to happen which is what the RetryStrategy is specifying.

Why does the developer need to make any choices here? The way I understand this is that for .NET 9+ the library should use CollectTrace4 with a rundown keyword and for .NET 8 and lower the library should use the existing commands with the rundown flag. I don't see where a choice from the developer that would come into play. This should be a pivot that can be deterministically detected and automatically handled withing the NETCore.Client library (so long as we allow profiles to be asked the correct question when getting the rundown keyword).

The only thing that I understand that needs to pivot is the CollectGC profile, which will want to report a non-zero keyword for runtimes that support rundown keywords and a zero keyword for runtimes that do not support rundown keywords (which indicates that this profile would not want any rundown at all, unless mixed with another profile/provider list that wants rundown). I spoke with @cshung briefly to confirm my understanding of the problem.

Here's a quick prototype of the idea that I'm trying convey: https://github.com/jander-msft/diagnostics/pull/1. It uses @cshung's idea to introduce the rundown keyword to the profiles, but calculates the keyword based on deterministic detection rather than using a custom specifiable retry strategy.
- The first commit refactors "request rundown" into getting a rundown keyword dynamically based on whether the target runtime supports rundown keyword vs requesting blanket rundown.
- The second commit adds CollectTracing4 and uses a very easily understood fallback algorithm.

These are what I think the benefits are:
- Removes the need for a state machine for performing a fallback algorithm.
- Removes the retry strategy notion.
- Minimizes the new code duplication between dotnet-trace and the Microsodt.Diagnostics.Monitoring library.
- All of the fallback algorithm is self-contained within the EventPipeSession class.

Is there some scenario that I'm missing that would not be covered by what I've written?

Reply:
If this concept is acceptable, I might tinker with it a little more to replace the GetRundownKeyword callback with property of a new type that reports both types of rundown information (the Boolean request for rundown and the rundown keyword). I think that'll further simplify the proposed solution.

Reply:
> Why does the developer need to make any choices here?

Pretend I am the developer of some tool that uses DiagnosticsClient and I have three different scenarios I want to implement using EventPipe:

- Scenario 1 - I want to collect some GC events and at the end I want a custom GC rundown event. That rundown isn't critical so I'd be willing to forego it on downlevel runtimes but I definitely do not want the standard rundown because that will be expensive and this profile is intended to be very lightweight.
- Scenario 2 - I want to collect some information about tiered compilation to show it to the user. The runtime has a CompilationLog rundown keyword that emits this info but it isn't on by default. For an older runtime I want my tool to emit an error saying that earlier runtimes aren't supported. Without this rundown data I'd have nothing to show the user and its pointless to have the user waiting for the tool to collect a trace that isn't configured to collect the relevant data.
- Scenario 3 - I want to collect some CPU sampling to show hot functions but I don't care about mapping to source lines. I need JitRundown to make this work. I also know that if I omit JittedMethodILToNativeMapKeyword from rundown the trace can be smaller. That would be a nice opportunistic optimization but if I am targetting .NET 8 and below I can accept having the trace be bigger to ensure the required data is there.

Scenario 1 is a case where I want to retry with rundown=false, scenario 2 is a case where I don't want to retry at all, and scenario 3 is a case where I want to retry with rundown=true. Although your implementation has IncludeDefaultRundownKeywords as part of the config, in scenario 3 I'd need that value false during the initial request but then I'd want it to be true during the retry. Overall I think I could make scenario 1 work, but I would struggle to implement scenario 2 and 3. In particular I'd need choices about:
- should the retry without custom keywords should be issued at all?
- if a retry without custom keywords is issued, should rundown be enabled or disabled?

If you want to hardcode a specific retry policy into System.Diagnostics.Monitoring.EventPipe I'd be fine with that. I don't want to hardcode the retry behavior in DiagnosticClient though.

Hope that helps clarify!

Reply:
Thank you for elaborating. I don't think this context of the possible future behavior of new profiles was explained at any point until now. I don't see a GH issue nor any discussion or listing of such a design or desire to enable those behaviors. All I have seen is the implementation detail in this PR that seemed to go beyond the needed capabilities of the current profiles.

> If you want to hardcode a specific retry policy into System.Diagnostics.Monitoring.EventPipe I'd be fine with that. I don't want to hardcode the retry behavior in DiagnosticClient though.

With that constraint, I retract my proposal.

Reply:
> I don't see a GH issue nor any discussion or listing of such a design or desire to enable those behaviors

I think you can assume as a standing design principle that DiagnosticClient wants to provide nearly the same flexibility as you would get writing the raw IPC messages. I don't think of it as supporting a concrete list of enumerated scenarios but rather a principle of transparently exposing the underlying IPC capabilities. Those scenarios just happened to be some examples to make the underlying capabilities more apparent and suggest why a future dev might care about being able to access them.

Whether you want to extend that flexibility/transparency up the stack in S.D.Monitoring.EventPipe I think is your call. It is not an explicit goal (right now at least) to implement scenarios (2) and (3) so you can apply your judgement whether you want to be flexible anticipating such scenarios might emerge in the future or you can have something more narrowly tailored to just the GC scenario which Andrew is pursuing right now.